File: C++/KthSmallestElementInABST.cc
// Given a binary search tree, write a function kthSmallest to find the kth smallest element in it.

// Note:
// You may assume k is always valid, 1 ≤ k ≤ BST's total elements.

class Solution {
    void inorderTraverse(TreeNode* root, int &k, int &val) {
        if (!root || k == 0) return;
        if (root->left) inorderTraverse(root->left, k, val);
        if (--k == 0) {
            val = root->val;
            return;
        }
        if (root->right) inorderTraverse(root->right, k, val);
    }

public:
    int kthSmallest(TreeNode* root, int k) {
        int val = 0;
        inorderTraverse(root, k, val);
        return val;
    }
};

File: Java/KthSmallestElementInABST.java
// Given a binary search tree, write a function kthSmallest to find the kth smallest element in it.

// Note:
// You may assume k is always valid, 1 ≤ k ≤ BST's total elements.

public class KthSmallestElementInABST {
    int i = 0;
    public int kthSmallest(TreeNode root, int k) {
        if (root == null) {
            return -1;
        }
        int l = kthSmallest(root.left, k);
        if (l != -1) return l;
        if (k == ++i) {
            return root.val;
        }
        int r = kthSmallest(root.right, k);
        if (r != -1) return r;
        return -1;
    }
}
